In [4]:
import pandas as pd
import numpy as np
import os

In [5]:
import warnings
warnings.filterwarnings('ignore')


# 1.The Problem
How to valuate a car in Australia
# 2.Causes of the analysis
This is to understand the car pricing in Australia

In [6]:
path = r'C:\Users\USER\Desktop\Cube\Data_Science\Project\Austrailian-car-sales'
os.chdir(path)
df = pd.read_excel('Australian-Vehicle-Prices.xlsx')
df

Brand    Year    Model               Car/Suv  \
0          Ssangyong  2022.0   Rexton  Sutherland Isuzu Ute   
1                 MG  2022.0      MG3             Hatchback   
2                BMW  2022.0     430I                 Coupe   
3      Mercedes-Benz  2011.0     E500                 Coupe   
4            Renault  2022.0   Arkana                   SUV   
...              ...     ...      ...                   ...   
16729         Toyota  2014.0  Alphard              SYC Cars   
16730  Mercedes-Benz  2012.0  S-Class        USED Dealer ad   
16731         Suzuki  2022.0   Vitara   Bremer Suzuki - New   
16732  Mercedes-Benz  2016.0   GLC250                   SUV   
16733  Mercedes-Benz  2021.0     C200                 Sedan   

                                           Title UsedOrNew Transmission  \
0           2022 Ssangyong Rexton Ultimate (awd)      DEMO    Automatic   
1      2022 MG MG3 Auto Excite (with Navigation)      USED    Automatic   
2                          2022 BMW 430I M Sport      USED    Automatic   
3               2011 Mercedes-Benz E500 Elegance      USED    Automatic   
4                     2022 Renault Arkana Intens      USED    Automatic   
...                                          ...       ...          ...   
16729                        2014 Toyota Alphard      USED    Automatic   
16730     2012 Mercedes-Benz S-Class CDI BlueTEC      USED    Automatic   
16731                    2022 Suzuki Vitara 1.6L      DEMO    Automatic   
16732                  2016 Mercedes-Benz GLC250      USED    Automatic   
16733                    2021 Mercedes-Benz C200      USED    Automatic   

             Engine DriveType  FuelType FuelConsumption Kilometres  \
0      4 cyl, 2.2 L       AWD    Diesel  8.7 L / 100 km       5595   
1      4 cyl, 1.5 L     Front   Premium  6.7 L / 100 km         16   
2        4 cyl, 2 L      Rear   Premium  6.6 L / 100 km       8472   
3      8 cyl, 5.5 L      Rear   Premium   11 L / 100 km     136517   
4      4 cyl, 1.3 L     Front  Unleaded    6 L / 100 km       1035   
...             ...       ...       ...             ...        ...   
16729             -     Front  Unleaded               -      66000   
16730             -      Rear    Diesel               -     117432   
16731  4 cyl, 1.6 L     Front  Unleaded  6.2 L / 100 km       5165   
16732    4 cyl, 2 L       AWD   Premium  7.2 L / 100 km      85525   
16733    4 cyl, 2 L      Rear  Unleaded    7 L / 100 km      31852   

        ColourExtInt             Location CylindersinEngine    BodyType  \
0      White / Black       Caringbah, NSW             4 cyl         SUV   
1      Black / Black       Brookvale, NSW             4 cyl   Hatchback   
2       Grey / White        Sylvania, NSW             4 cyl       Coupe   
3      White / Brown    Mount Druitt, NSW             8 cyl       Coupe   
4       Grey / Black     Castle Hill, NSW             4 cyl         SUV   
...              ...                  ...               ...         ...   
16729       Grey / -         Rocklea, QLD                 -  Commercial   
16730     Silver / -  Port Melbourne, VIC                 -         NaN   
16731     Yellow / -         Dinmore, QLD             4 cyl         SUV   
16732   Grey / Black          Albion, QLD             4 cyl         SUV   
16733       Gold / -          Yatala, QLD             4 cyl       Sedan   

          Doors     Seats   Price  
0       4 Doors   7 Seats   51990  
1       5 Doors   5 Seats   19990  
2       2 Doors   4 Seats  108988  
3       2 Doors   4 Seats   32990  
4       4 Doors   5 Seats   34990  
...         ...       ...     ...  
16729   7 Seats       NaN   29500  
16730       NaN       NaN   39888  
16731   4 Doors   5 Seats   35280  
16732   4 Doors   5 Seats   41888  
16733   4 Doors   5 Seats   65888  

[16734 rows x 19 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16734 entries, 0 to 16733
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Brand              16733 non-null  object 
 1   Year               16733 non-null  float64
 2   Model              16733 non-null  object 
 3   Car/Suv            16706 non-null  object 
 4   Title              16733 non-null  object 
 5   UsedOrNew          16733 non-null  object 
 6   Transmission       16733 non-null  object 
 7   Engine             16733 non-null  object 
 8   DriveType          16733 non-null  object 
 9   FuelType           16733 non-null  object 
 10  FuelConsumption    16733 non-null  object 
 11  Kilometres         16733 non-null  object 
 12  ColourExtInt       16733 non-null  object 
 13  Location           16284 non-null  object 
 14  CylindersinEngine  16733 non-null  object 
 15  BodyType           16452 non-null  object 
 16  Doors              151

In [8]:
df.columns

Index(['Brand', 'Year', 'Model', 'Car/Suv', 'Title', 'UsedOrNew',
       'Transmission', 'Engine', 'DriveType', 'FuelType', 'FuelConsumption',
       'Kilometres', 'ColourExtInt', 'Location', 'CylindersinEngine',
       'BodyType', 'Doors', 'Seats', 'Price'],
      dtype='object')

In [9]:
df.isnull().sum()

Brand                   1
Year                    1
Model                   1
Car/Suv                28
Title                   1
UsedOrNew               1
Transmission            1
Engine                  1
DriveType               1
FuelType                1
FuelConsumption         1
Kilometres              1
ColourExtInt            1
Location              450
CylindersinEngine       1
BodyType              282
Doors                1604
Seats                1705
Price                   3
dtype: int64

In [10]:
df.describe()

Year
count  16733.000000
mean    2016.229248
std        5.247705
min     1940.000000
25%     2013.000000
50%     2017.000000
75%     2020.000000
max     2023.000000

In [11]:
df.duplicated().sum()
#No duplicates in the dataset

np.int64(0)

In [12]:
print(df['Brand'].value_counts())
print(df['Year'].value_counts())
print(df['Model'].value_counts())
print(df['DriveType'].value_counts())
print(df['Location'].value_counts())
print(df['UsedOrNew'].value_counts())

Brand
Toyota      2784
Hyundai     1239
Mazda       1179
Holden      1087
Ford        1055
            ... 
Daewoo         1
Tata           1
Polestar       1
Datsun         1
Packard        1
Name: count, Length: 76, dtype: int64
Year
2018.0    1599
2019.0    1480
2017.0    1350
2022.0    1328
2023.0    1319
2020.0    1135
2016.0    1124
2021.0    1098
2015.0    1083
2014.0     876
2013.0     851
2012.0     696
2011.0     519
2010.0     454
2009.0     378
2008.0     325
2007.0     270
2006.0     220
2005.0     173
2004.0     117
2003.0      74
2002.0      64
2000.0      44
2001.0      37
1998.0      30
1999.0      29
1997.0       9
1994.0       9
1996.0       8
1993.0       7
1989.0       5
1990.0       5
1995.0       2
1985.0       2
1992.0       2
1981.0       2
1986.0       1
1991.0       1
1959.0       1
1970.0       1
1984.0       1
1975.0       1
1979.0       1
1978.0       1
1940.0       1
Name: count, dtype: int64
Model
Hilux          430
Corolla        405
Ranger         398


### 🔧 Brand Frequency (`Brand`)
- **Top Brands:**
  - Toyota: 2,784
  - Hyundai: 1,239
- **Total Unique Brands:** 76  
- **Rare Brands:** Daewoo, Tata, Polestar, Datsun, Packard (each with 1 listing)

### 📆 Year of Manufacture (`Year`)
- **Most Common Years:**  
  - 2018 (1,599), 2019 (1,480), 2017 (1,350), 2022 (1,328), 2023 (1,319)
- **Trend:** Majority of vehicles listed are from 2015–2023.
- **Oldest Vehicle Year:** 1940 (1 listing)

### 🚘 Car Model (`Model`)
- **Most Frequent Models:**  
  - Hilux (430), Corolla (405), Ranger (398), Landcruiser (370), I30 (366)
- **Total Unique Models:** 781  
- **High model diversity** across the dataset.

### 🚗 Drive Type (`DriveType`)
- **Distribution:**
  - Front: 6,978
  - 4WD: 3,143
  - AWD: 3,113
  - Rear: 2,350
  - Other: 1,149

### 📍 Location (`Location`)
- **Top Locations (most listings):**
  - Minchinbury, NSW: 544
  - Blacktown, NSW: 295
  - Liverpool, NSW: 254
- **Total Unique Locations:** 618  

### 🏷️ Vehicle Condition (`UsedOrNew`)
- **USED:** 14,994
- **NEW:** 1,227
- **DEMO:** 512
- **Observation:** Used cars dominate the market (~85% of listings).

# 3.Data cleaning
This is dropping unnecessary columns and imputing values using mean,median and mode

In [13]:
df.isnull().sum()

Brand                   1
Year                    1
Model                   1
Car/Suv                28
Title                   1
UsedOrNew               1
Transmission            1
Engine                  1
DriveType               1
FuelType                1
FuelConsumption         1
Kilometres              1
ColourExtInt            1
Location              450
CylindersinEngine       1
BodyType              282
Doors                1604
Seats                1705
Price                   3
dtype: int64

In [14]:
#Dropping unnecessary columns
df.drop(columns=['Car/Suv','Title', 'Doors'],inplace=True)
# Car/Suv column is not useful for our analysis as it is a duplicate of the Body Type column
# Title column is not useful for our analysis as it has been extacted into Brand, Year and Model columns
# Doors column is not useful for our analysis as Seats column is more informative about the car's capacity

In [15]:
df.keys()

Index(['Brand', 'Year', 'Model', 'UsedOrNew', 'Transmission', 'Engine',
       'DriveType', 'FuelType', 'FuelConsumption', 'Kilometres',
       'ColourExtInt', 'Location', 'CylindersinEngine', 'BodyType', 'Seats',
       'Price'],
      dtype='object')

In [16]:
df.sample(5)

Brand    Year        Model UsedOrNew Transmission        Engine  \
13736   Toyota  2022.0        Hilux      USED    Automatic  4 cyl, 2.8 L   
13632   Toyota  2022.0  Landcruiser      USED    Automatic  4 cyl, 2.8 L   
3816      Land  2015.0        Rover      USED    Automatic    8 cyl, 5 L   
15727   Toyota  2005.0  Landcruiser      USED       Manual  6 cyl, 4.2 L   
1335   Hyundai  2020.0          I30      USED    Automatic    4 cyl, 2 L   

      DriveType  FuelType  FuelConsumption Kilometres   ColourExtInt  \
13736       4WD    Diesel     8 L / 100 km      31626      White / -   
13632       4WD    Diesel   7.9 L / 100 km      19569      Brown / -   
3816        AWD   Premium  13.8 L / 100 km      66250  Black / Black   
15727       4WD    Diesel     0 L / 100 km     361771      White / -   
1335      Front  Unleaded   7.4 L / 100 km      47705       Grey / -   

               Location CylindersinEngine    BodyType     Seats   Price  
13736      Jandakot, WA             4 cyl  Ute / Tray   5 Seats   71984  
13632     Pinelands, NT             4 cyl         SUV   7 Seats   89990  
3816         Glebe, NSW             8 cyl         SUV   5 Seats  109950  
15727    Granville, NSW             6 cyl         SUV   8 Seats   34999  
1335   Pendle Hill, NSW             4 cyl   Hatchback   5 Seats   22950

# 3.1 Handling missing values

In [17]:
df.isnull().sum()/df.isnull().count().sort_values()*100 
#Getting the missing values per coulmn in percentages

Brand                 0.005976
Year                  0.005976
Model                 0.005976
UsedOrNew             0.005976
Transmission          0.005976
Engine                0.005976
DriveType             0.005976
FuelType              0.005976
FuelConsumption       0.005976
Kilometres            0.005976
ColourExtInt          0.005976
Location              2.689136
CylindersinEngine     0.005976
BodyType              1.685192
Seats                10.188837
Price                 0.017928
dtype: float64

In [18]:
# Dropping the rows with missing values

col_with_missing = ['Brand', 'Year', 'Model', 'CylindersinEngine', 'Transmission','Engine', 'DriveType','FuelType','FuelConsumption', 'UsedOrNew', 'ColourExtInt','Kilometres','Price', 'BodyType', 'Seats', 'Location']
df = df.dropna(subset=col_with_missing, how='any')

In [19]:
df.isna().sum()

Brand                0
Year                 0
Model                0
UsedOrNew            0
Transmission         0
Engine               0
DriveType            0
FuelType             0
FuelConsumption      0
Kilometres           0
ColourExtInt         0
Location             0
CylindersinEngine    0
BodyType             0
Seats                0
Price                0
dtype: int64

### Now our dataframe has no missing values 

# 3.2 Feature Engineering

In [20]:
# Split into two columns (City and State)
df[['City', 'State']] = df['Location'].str.split(',', expand=True)

df['City'] = df['City'].str.strip()
df['State'] = df['State'].str.strip()

# Drop the original 'Location' column 
df = df.drop('Location', axis=1)

In [21]:
df['City'].value_counts()

City
Minchinbury         503
Liverpool           251
Blacktown           236
Hoppers Crossing    219
Ringwood            208
                   ... 
Norwood               1
Sale                  1
Casino                1
Beaudesert            1
East Toowoomba        1
Name: count, Length: 582, dtype: int64

In [22]:
df['CylindersinEngine'].value_counts()

CylindersinEngine
4 cyl     11203
6 cyl      1982
8 cyl       585
5 cyl       399
3 cyl       309
0 L          90
12 cyl       11
2 cyl         4
10 cyl        3
Name: count, dtype: int64

In [23]:
# Extracting the fuel consumption for every 100km from the FuelConsumption column
df['FuelConsumption'] = df['FuelConsumption'].apply(lambda x: float(str(x).split()[0]))

# Renaming the ule consumption column to FuelConsumptionPer100km
df.rename(columns={'FuelConsumption': 'FuelConsumptionPer100km'}, inplace=True)

In [24]:
#Extracting the engine size from the Engine column
df['Engine'] = df['Engine'].str.split(',').str[1].str.strip('L')

#Conert to floats
df['Engine'] = pd.to_numeric(df['Engine'], errors='coerce') 

# Renaming the Engine column to EngineType
df.rename(columns={'Engine': 'EngineSize'}, inplace=True)

In [25]:
# Extracting the exterior color from the colourExtInt column since Int has some missing values
df['ColourExtInt'] = df['ColourExtInt'].str.split('/').str[0].str.strip()

# Renaming the ColourExtInt column to ExteriorColour
df.rename(columns={'ColourExtInt': 'ExteriorColour'}, inplace=True)

In [26]:
# Converting  the CylindersinEngine to floats
df['CylindersinEngine'] = df['CylindersinEngine'].apply(lambda x: int(str(x).split()[0]) if isinstance(x, str) else x)

# Making the Seats columnto be numeric
df['Seats'] = df['Seats'].apply(lambda x: int(str(x).split()[0]) if isinstance(x, str) else x)

# Converting year into datetime
df['Year'] = pd.to_datetime(df['Year'], format='%Y').dt.year

The engine size now has missing values which we will drop

In [27]:
#Dropping null values in engine size
df = df.dropna(subset=['EngineSize'], how='any')

In [28]:
# saving the cleaned dataset
df.to_csv('Australian_Vehicle_Prices_Cleaned.csv', index=False) 

In [127]:
df.head(10)

Brand  Year   Model UsedOrNew Transmission  EngineSize DriveType  \
0       Ssangyong  2022  Rexton      DEMO    Automatic         2.2       AWD   
1              MG  2022     MG3      USED    Automatic         1.5     Front   
2             BMW  2022    430I      USED    Automatic         2.0      Rear   
3   Mercedes-Benz  2011    E500      USED    Automatic         5.5      Rear   
4         Renault  2022  Arkana      USED    Automatic         1.3     Front   
6            Land  2017   Rover      USED    Automatic         2.0       AWD   
7          Nissan  2000  Pulsar      USED    Automatic         1.6     Front   
8          Toyota  2013      86      USED    Automatic         2.0      Rear   
9           Honda  2014    Jazz      USED    Automatic         1.3     Front   
11         Toyota  2018   HiAce      USED    Automatic         2.7      Rear   

    FuelType  FuelConsumptionPer100km Kilometres ExteriorColour  \
0     Diesel                      8.7       5595          White   
1    Premium                      6.7         16          Black   
2    Premium                      6.6       8472           Grey   
3    Premium                     11.0     136517          White   
4   Unleaded                      6.0       1035           Grey   
6     Diesel                      5.1      67662          White   
7   Unleaded                      8.0     300539            Red   
8    Premium                      7.1      82012          Black   
9     Hybrid                      4.5      38229           Blue   
11  Unleaded                      9.8      16324          White   

    CylindersinEngine    BodyType  Seats   Price            City State  
0                   4         SUV      7   51990       Caringbah   NSW  
1                   4   Hatchback      5   19990       Brookvale   NSW  
2                   4       Coupe      4  108988        Sylvania   NSW  
3                   8       Coupe      4   32990    Mount Druitt   NSW  
4                   4         SUV      5   34990     Castle Hill   NSW  
6                   4         SUV      4   62280       Blacktown   NSW  
7                   4   Hatchback      5    2995  Wentworthville   NSW  
8                   4       Coupe      4   24888   Mcgraths Hill   NSW  
9                   4   Hatchback      5   17900        Lidcombe   NSW  
11                  4  Commercial      2   41999        Lidcombe   NSW

In [128]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14495 entries, 0 to 16733
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Brand                    14495 non-null  object 
 1   Year                     14495 non-null  int32  
 2   Model                    14495 non-null  object 
 3   UsedOrNew                14495 non-null  object 
 4   Transmission             14495 non-null  object 
 5   EngineSize               14495 non-null  float64
 6   DriveType                14495 non-null  object 
 7   FuelType                 14495 non-null  object 
 8   FuelConsumptionPer100km  14495 non-null  float64
 9   Kilometres               14495 non-null  object 
 10  ExteriorColour           14495 non-null  object 
 11  CylindersinEngine        14495 non-null  int64  
 12  BodyType                 14495 non-null  object 
 13  Seats                    14495 non-null  int64  
 14  Price                    14

After extracting new features from columns and solving missing values we have a dataset with 14495 values

# 4. Visualization


In [135]:
import plotly.express as px

# Visualizing car prices against the year of manufacture
fig = px.histogram(df, x='Year', y='Price', color='Brand', title='Car Prices by Year of Manufacture',
                   labels={'Year': 'Year of Manufacture', 'Price': 'Price (in AUD)', 'Brand': 'Car Brand'},
                   nbins=20, histnorm='percent', barmode='overlay')
fig.update_layout(xaxis_title='Year of Manufacture',
                  yaxis_title='Price (in AUD)',
                  legend_title='Car Brand',
                  template='plotly_white',
                  xaxis=dict(tickmode='linear', dtick=1),
                  yaxis=dict(tickformat=',.0f'))  # Format y-axis as integer
fig.show()
